stdlib/json.gr

```diff
@@ -0,0 +1,2157 @@
+    emitUTF16EscapeSequence(lowSurrogate, buffer)
+  }
+}
+
+let emitEscapedCodePoint = (codePoint: Number, buffer: Buffer.Buffer) => {
+  match (codePoint) {
+    0x0008 => { // backspace
+      Buffer.addChar('\\', buffer)
+      Buffer.addChar('b', buffer)
+    },
+    0x0009 => { // tab
+      Buffer.addChar('\\', buffer)
+      Buffer.addChar('t', buffer)
+    },
+    0x000A => { // line feed
+      Buffer.addChar('\\', buffer)
+      Buffer.addChar('n', buffer)
+    },
+    0x000C => { // form feed
+      Buffer.addChar('\\', buffer)
+      Buffer.addChar('f', buffer)
+    },
+    0x000D => { // carriage return
+      Buffer.addChar('\\', buffer)
+      Buffer.addChar('r', buffer)
+    },
+    0x0022 => { // quotation mark
+      Buffer.addChar('\\', buffer)
+      Buffer.addChar('"', buffer)
+    },
+    0x005C => { // backslash or "Reverse Solidus"
+      Buffer.addChar('\\', buffer)
+      Buffer.addChar('\\', buffer)
+    },
+    _ => {
+      emitEscapedUnicodeSequence(codePoint, buffer)
+    },
+  }
+}
+
+let printNull = (buffer: Buffer.Buffer) => Buffer.addString("null", buffer)
+
+let printBool = (b: Bool, buffer: Buffer.Buffer) => {
+  if (b) {
+    Buffer.addString("true", buffer)
+  } else {
+    Buffer.addString("false", buffer)
+  }
+}
+
+exception UnknownNumberTag
+
+@unsafe
+let printNumberWasmI32 = (value: WasmI32, buffer: Buffer.Buffer) => {
+  let s = NumberUtils.itoa32(value, 10n)
+  Buffer.addString(s, buffer)
+}
+
+@unsafe
+let printNumberWasmI64 = (value: WasmI64, buffer: Buffer.Buffer) => {
+  let s = NumberUtils.itoa64(value, 10n)
+  Buffer.addString(s, buffer)
+}
+
+@unsafe
+let isFinite = (value: WasmF64) => {
+  from WasmF64 use { (==), (-) }
+  value - value == 0.0W
+}
+
+@unsafe
+let isNaN = (value: WasmF64) => {
+  from WasmF64 use { (!=) }
+  value != value
+}
+
+@unsafe
+let printNumberWasmF64 = (value: WasmF64, buffer: Buffer.Buffer) => {
+  if (isFinite(value)) {
+    let s = NumberUtils.dtoa(value)
+    Buffer.addString(s, buffer)
+    None
+  } else {
+    from WasmF64 use { (<) }
+    // JSON standard doesn't allow NaN or infinite values in numbers,
+    // but WASM f64 (IEEE 754-2008), as well as Grain's number types do
+    // (Float64 as well as Number). This is the only reason that the
+    // formatting needs to return a Result and not just a String
+    // directly. Other possible choices were to throw exceptions or to
+    // coninue formatting without representing these values correctly
+    // (like JavaScript's JSON.stringify).
+    if (isNaN(value)) {
+      Some(InvalidNumber("NaN is not allowed in JSONNumber"))
+    } else if (value < 0.0W) {
+      Some(InvalidNumber("-Infinity is not allowed in JSONNumber"))
+    } else {
+      Some(InvalidNumber("Infinity is not allowed in JSONNumber"))
+    }
+  }
+}
+
+@unsafe
+let printNumber = (value: Number, buffer: Buffer.Buffer) => {
+  from WasmI32 use { (&), (==), (!=), (<<), (>>) }
+
+  let ptr = WasmI32.fromGrain(value)
```

Comment:
For good measure, let's add an `ignore(value)` to the end of this function.

Reply:
By end you mean before the return values right?

Reply:
I meant at the very end; first assigning the result of the `if` to a value so you can return that.